File: examples/deno/hello.ts
/**
 * Hello, resource — in TypeScript.
 *
 * The smallest ikigai peer there is: one function, one `endpoint(...)`, one socket. Run
 * it and a Rust host can mount it; `--mount urn:ts:=<socket>` in the chapter is the
 * binding.
 *
 *     deno run -A examples/deno/hello.ts /tmp/ts-hello.sock
 *
 * Modeled on ikigai-deno's `examples/demo.ts` (read-only for this book); kept here so
 * the book builds from a clone of this repository alone. Imports ikigai-deno by path
 * until it is on JSR — set `IKIGAI_DENO` to your checkout, or edit the import.
 */

import { endpoint, Server } from "../../../ikigai-deno/src/serve.ts";

const XSD_STRING = "http://www.w3.org/2001/XMLSchema#string";

// ANCHOR: endpoint
// The description is explicit here — `args:` is the ArgSpec list, stated the way the
// Rust chapter stated it — and the handler receives the named arguments. (ikigai-deno's
// `./zod` entry point derives the same list from a schema; this file stays
// zero-dependency on purpose.) `cacheable: true` is the same claim `.cacheable()` makes
// in Rust, and the mounting kernel is what honors it.
export const hello = endpoint("urn:ts:hello", {
  summary: "Greet someone",
  args: [{
    name: "who",
    required: true,
    summary: "the name to greet",
    class: XSD_STRING,
  }],
  cacheable: true,
}, ({ who }) => `Hello, ${who}!`);

export const shout = endpoint("urn:ts:shout", {
  summary: "Uppercase a string, loudly",
  args: [{
    name: "text",
    required: true,
    summary: "the text to shout",
    class: XSD_STRING,
  }],
  cacheable: true,
}, ({ text }) => `${String(text).toUpperCase()}!`);
// ANCHOR_END: endpoint

// ANCHOR: serve
if (import.meta.main) {
  const path = Deno.args[0] ?? "/tmp/ts-hello.sock";
  console.error(`serving urn:ts:hello and urn:ts:shout on ${path}`);
  const server = new Server([hello, shout], path);
  Deno.addSignalListener("SIGINT", () => server.shutdown());
  await server.serve(); // blocks; speaks the wire protocol
}
// ANCHOR_END: serve

File: examples/python/hello.py
"""Hello, resource — in Python.

The smallest ikigai peer there is: one function, one decorator, one socket. Run it and
a Rust host can mount it; `--mount urn:py:=<socket>` in the chapter is the binding.

    python3 examples/python/hello.py /tmp/py-hello.sock

Modeled on ikigai-python's ``examples/endpoints.py`` (read-only for this book); kept
here so the book builds from a clone of this repository alone. Needs ``ikigai-python``
installed (``pip install <path to ikigai-python>``); it has no dependencies of its own.
"""

from __future__ import annotations

import sys
from typing import Annotated

from ikigai import endpoint, serve

# ANCHOR: endpoint
# The decorator is the description, and the signature is the contract. `who: str`
# becomes a required input of class xsd:string; the `Annotated` text becomes its
# summary; `cacheable=True` is the same claim `.cacheable()` makes in Rust — a pure
# function of its declared inputs — and it is the mounting kernel, not this process,
# that will honor it.
@endpoint("urn:py:hello", summary="Greet someone", cacheable=True)
def hello(who: Annotated[str, "the name to greet"]) -> str:
    return f"Hello, {who}!"


# A second one, so `list` has something to show and a pipe has something to feed.
@endpoint("urn:py:shout", summary="Uppercase a string, loudly", cacheable=True)
def shout(text: Annotated[str, "the text to shout"]) -> str:
    return text.upper() + "!"
# ANCHOR_END: endpoint


# ANCHOR: serve
if __name__ == "__main__":
    path = sys.argv[1] if len(sys.argv) > 1 else "/tmp/py-hello.sock"
    print(f"serving urn:py:hello and urn:py:shout on {path}", file=sys.stderr)
    serve([hello, shout], path)  # blocks; speaks the wire protocol
# ANCHOR_END: serve
